### ISSUE TITLE
Simplify/remove base class methods

### ISSUE THREAD

**Issue by dschult** [networkx/networkx#2548]
We've often talked about removing methods from the base classes. V2.0 is one opportunity to do that. Which of these options should we do?

1. **Move self-loop methods to functions**
We have already moved ```add_path``` and friends to functions.  What about the methods reporting self-loops?:  ```selfloop_edges```, ```number_of_selfloops```, and ```nodes_with_selfloops```.

2. **Combine ```G.node``` with ```G.nodes``` and ```G.edge``` with ```G.edges```**
since they do essentially the same thing now. The plural versions allow both set operations and dict-like lookup with ```.data``` added to ```keys/values/items```. The singular versions don't provide set operations or the ```data``` method.

3. **Remove ```neighbors/predecessors/succesors``` methods in favor of set-dict objects ```adj/succ/pred```**
This might be more user friendly if we change ```G.adj``` to ```G.nbrs```. Then we can also programmatically tell the difference between v1.x and v2.x.

4. **Move ```subgraph/reverse/to_directed/to_undirected``` to functions**

**Comment by Erotemic** [comment]
IMO:

1. It seems like the self-loop functions are too specific to belong to the base graph class. I think it would be better to call them via `nx.selfloop_edges`. 

2. It took me by surprise when `G.edge` and `G.node` were changed. It is probably better to remove them entirely. They're already broken from the perspective of a `1.x` user, so may as well just throw the `AttributeError`. (On a side note, I didn't know about `G.edges.data()`, I have to say I like it). 

3. I actually do like having `neighbors/predecessors/succesors`, and I'd vote for keeping them. As for telling the difference between 1.x and 2.x, isn't that what `nx.__version__` is for?

4. I'm opposed to moving `G.subgraph` to its own function. It feels like a very fundamental operation for me and it applies to Graphs, DiGraphs, and MultiGraphs. Removing this would be like removing `set().issubset`.  As for `reverse`, I sort of like it as a method, but its only valid for Digraphs, so I'm not against moving it to a function. I'm similarly ambivalent about the `to_directed/to_undirected` API. However, I'll reiterate my strong feelings on keeping `G.subgraph`.

**Comment by dschult** [comment]
Thanks for this feedback. It is very helpful.  I'm leaning the way you mention on each topic.

Could you expand on what you like about ```neighbors/predecessors/successors```? It basically comes down to ```G.neighbors(n)``` vs ```list(G.adj[n])```.  Is it the abbreviation, the need for list, the head-bending of a dict turning into a list?  Said the opposite way: is it the full name, the simplicity of one name without post-processing via list, the simplicity of not having to think about an adjacency dict?

**Comment by Erotemic** [comment]
I think its the abbreviation / full name. 

What I like about neighbors/predecessors/successors comes down to semantics. When I'm coding I typically do it with IPython, and I'll type `nx.` and then try a few keywords plus tab to find something that looks like the function I'm looking for. Neighbors, predecessors, and successors are very intuitive, I think using them makes the code more readable as well. 

To me, this:
``` python
for node in G.neighbors(n):
    dostuff(node)
```

is more clear than this:
```python
for node in G.adj[n]:
    dostuff(node)
```

I've used NetworkX in a labs as a TA many times, and I always urge them to use networkx as their basic data structure (after they've already implemented their own adjacency list data structure). For instance when a student is struggling to implement DFS, I walk them through it. "Now, you need to get the neighbors of the current node. How might you do that?"  ... they struggle ... "Well what about the neighbors function?" ... and then things start to click for them. This is anecdotal, but I feel like the nice semantics of networkx are one of its strongest selling points. 

There is something to be said about confusion that might arise because `G.neighbors(n)` and `list(G.adj[n])` doing the exact same thing, so my position isn't staunch.

**Comment by dschult** [comment]
This is also very helpful.  Thanks...

Questions (I agree that semantics are important:

 - ```G.nbrs[n]```      vs ```G.adj[n]```      (adjacency is the more "graph theory" term but neighbors is nice.)

 - ```nbunch``` and ```ebunch```     vs    ```nodes``` and ```edges```    (These only get used as arguments to methods.)

**Comment by Erotemic** [comment]
I think `G.adj[n]` is better than `G.nbrs[n]` because once you start abbreviating its a sign that things are getting more technical. (Also I've always disliked nbrs as an abbreviation for neighbors). NetworkX is really nice for addressing graph theory related tasks. From a teaching "Intro to Algorithms" perspective,  having the `G.adj` structure makes a lot of sense because it shows the students how methods like `G.neighbors(n)` are working under the hood. 

So, I would always prefer the graph theory term over some other term when it comes to the underlying data structures. The more semantic and un-abbreviated names are better left for convenience methods/functions. 

Anecdote: When I was formalizing my understanding of graph theory I actually learned about adjacency matrices by looking at the source code for G.neighbors (or maybe a similar function, it was awhile ago), but the point is the semantic name led me to the underlying graph theory term.

 On `nbunch` vs `ebunch`, I'm pretty ambivalent. If I recall correctly, many functions that take a `*bunch` argument can either accept a single node/edge or a sequence of nodes/edges? That is an argument for leaving the `*bunch` name, so the plural nodes/edges always refer to a list of items. On the other hand `nodes/edges` are much more intuitive when specifying keyword arguments. Either way its extremely minor because argument names are so rarely exposed. If I had to vote I'd probably just keep `*bunch` due to the principle of least action, but this vote should be considered with an effective weight of ε.
